libafl/src/events/mod.rs

```diff
@@ -452,13 +448,26 @@ pub trait EventFirer: UsesState {
 }
 
 /// [`ProgressReporter`] report progress to the broker.
-pub trait ProgressReporter: EventFirer
-where
-    Self::State: HasMetadata + HasExecutions + HasLastReportTime,
-{
+pub trait ProgressReporter: UsesState {
     /// Given the last time, if `monitor_timeout` seconds passed, send off an info/monitor/heartbeat message to the broker.
     /// Returns the new `last` time (so the old one, unless `monitor_timeout` time has passed and monitor have been sent)
     /// Will return an [`Error`], if the stats could not be sent.
+    fn maybe_report_progress(
+        &mut self,
+        state: &mut Self::State,
+        monitor_timeout: Duration,
+    ) -> Result<(), Error>;
+
+    /// Send off an info/monitor/heartbeat message to the broker.
+    /// Will return an [`Error`], if the stats could not be sent.
+    fn report_progress(&mut self, state: &mut Self::State) -> Result<(), Error>;
+}
+
+impl<EF> ProgressReporter for EF
```

Comment:
Doesn't this mean downstream ProgressReporters cannot implement their own behavior anymore?

Reply:
Sure, but every progress reporter has used the default impls. If we want, we can also split this into `impl DefaultProgressReporter` or even just generic functions, e.g. `default_report_progress`

Reply:
Point is, right now overriding some default behavior "just works" after this it won't work anymore

Reply:
For a library user this makes a difference. And if we want to add new methods to a trait it matters as well

Reply:
Certainly. We have to make these decisions in an opinionated manner, on a case-by-case basis.

From my perspective, there is very rarely any reason to actually implement alternative progress reporters. We have two real options to reduce generics:
 - Any option we choose that doesn't enforce a specific implementation would remove generics from ~everything trying to write anything involving a progress reporter at the cost of ~10 lines of code per implementation of progress reporter.
 - We enforce a specific implementation for everyone. We invalidate downstream implementions of ProgressReporter, but make the LoC cost for all implementations 0.